File: dashboard/electron/renderer/overlay/observation.js
// Truthful tri-state values for review evidence — the JS twin of core/observation.py.
// The wire literals are identical on both sides, so a renderer can never disagree
// with the module that produced the value.
//
// A DRS check has THREE outcomes, never two:
//   observed_true   the system looked and the thing was there
//   observed_false  the system looked and the thing was not there
//   not_observed    the system could not look at all
//
// Rule for consumers: never test one of these for truthiness. Compare to a member,
// or ask isKnown() first. UNKNOWN must render as its OWN visual state — never as
// the negative one. (A placeholder "dislodged" once reached the evidence frame as a
// red "BAILS DISLODGED"; naively removing it would have printed "BAILS INTACT",
// which is equally false.)

export const Observation = Object.freeze({
  TRUE: "observed_true",
  FALSE: "observed_false",
  UNKNOWN: "not_observed",
});

// Accepts a wire string or a legacy boolean/null. Anything unrecognised is UNKNOWN
// — an unreadable value is not evidence.
export function observation(value) {
  if (value === true) return Observation.TRUE;
  if (value === false) return Observation.FALSE;
  if (value === Observation.TRUE || value === Observation.FALSE) return value;
  return Observation.UNKNOWN;
}

export const isKnown = (value) => observation(value) !== Observation.UNKNOWN;

// Operator wording. `unknown` states what the camera saw rather than blaming the
// detector — "not detected" reads as a model failure to an umpire.
export function observationLabel(value, yes, no, unknown = "Not observed") {
  const o = observation(value);
  return o === Observation.TRUE ? yes : o === Observation.FALSE ? no : unknown;
}

export const BailsState = Object.freeze({
  DISLODGED: "dislodged",
  INTACT: "intact",
  NOT_OBSERVED: "not_observed",
});

// Legacy payloads used null for "no detector". That is NOT_OBSERVED, and must
// never be read as INTACT.
export function bailsState(value) {
  const v = String(value ?? "").toLowerCase();
  return v === BailsState.DISLODGED || v === BailsState.INTACT ? v : BailsState.NOT_OBSERVED;
}

export function bailsLabel(value) {
  const s = bailsState(value);
  return s === BailsState.DISLODGED ? "Dislodged" : s === BailsState.INTACT ? "Intact" : "Not observed";
}

// null when unobserved — callers must not treat that as "not broken".
export function wicketBroken(value) {
  const s = bailsState(value);
  return s === BailsState.NOT_OBSERVED ? null : s === BailsState.DISLODGED;
}

/* ===================== FRAME IDENTITY — twin of core/frame_ref.py =====================
   Three different integers named `frame_id` circulate in this system: per-camera
   capture counters, replay-clip positions, and (historically) synthesised array
   indices. Comparing across them draws an overlay on the WRONG MOMENT, which is
   misleading evidence rather than a cosmetic bug. `timestamp_ms` is the only value
   that means the same thing everywhere, so it is the cross-surface join key. */

export const FrameSpace = Object.freeze({
  CAPTURE: "capture",   // per-camera VideoFrame.frame_id counter
  CLIP: "clip",         // 0..total-1 within one frozen replay window
});

export class FrameSpaceMismatch extends Error {}

export function captureFrame(index, timestampMs = null, cameraId = null) {
  return {
    space: FrameSpace.CAPTURE,
    index: Number(index),
    timestamp_ms: timestampMs == null ? null : Number(timestampMs),
    source: cameraId == null ? null : `camera:${cameraId}`,
  };
}

export function clipFrame(index, timestampMs = null, window = null) {
  return {
    space: FrameSpace.CLIP,
    index: Number(index),
    timestamp_ms: timestampMs == null ? null : Number(timestampMs),
    source: window ?? null,
  };
}

// Comparable only within ONE space AND one source: camera 0 frame 194 and
// camera 1 frame 194 are different instants.
export function comparableWith(a, b) {
  return Boolean(a && b) && a.space === b.space && (a.source ?? null) === (b.source ?? null);
}

// Throws rather than returning false for an incomparable pair — a quiet false
// reads as "a different moment" when the honest answer is "unanswerable".
export function isSameMoment(a, b) {
  if (!comparableWith(a, b)) {
    throw new FrameSpaceMismatch(
      `cannot compare ${a && a.space}/${a && a.source} with ${b && b.space}/${b && b.source}` +
      " — join on timestamp_ms instead");
  }
  return a.index === b.index;
}

// The frame an overlay should draw for, given where the umpire is in the clip.
// Returns the tracked point whose timestamp is nearest — never an index match,
// because the track is in capture space and the scrubber is in clip space.
export function trackPointAt(track, timestampMs) {
  if (!Array.isArray(track) || !track.length || timestampMs == null) return null;
  let best = null, bestGap = Infinity;
  for (const point of track) {
    const ts = point && point.frame && point.frame.timestamp_ms;
    if (ts == null) continue;
    const gap = Math.abs(ts - timestampMs);
    if (gap < bestGap) { bestGap = gap; best = point; }
  }
  return best;
}

File: core/observation.py
"""Truthful tri-state values for review evidence.

A DRS check has THREE outcomes, never two:

    observed_true    the system looked and the thing was there
    observed_false   the system looked and the thing was not there
    not_observed     the system could not look at all

Collapsing that into a boolean is how a placeholder ``bails_status = "dislodged"``
ended up painted onto an evidence frame as a red "BAILS DISLODGED" for every
run-out review — asserting to an umpire something no detector had measured. The
failure mode is subtle in the other direction too: ``None`` looks like a safe
"unknown" until a consumer writes ``if analysis.get("gloves_detected")`` and
silently reads it as "no gloves".

So the tri-state is explicit, and every value is a plain string on the wire. The
enums give type safety inside Python; JSON stays human-readable; and the
JavaScript renderers switch on the identical literals (see
``dashboard/electron/renderer/overlay/observation.js``, the twin of this module).

Rule for consumers: never test one of these values for truthiness. Compare to a
member, or ask ``is_known`` first. ``UNKNOWN`` must render as its own visual
state — never as the negative one.
"""

from __future__ import annotations

from enum import Enum


class Observation(str, Enum):
    """A yes/no question the system may or may not have been able to answer."""

    TRUE = "observed_true"
    FALSE = "observed_false"
    UNKNOWN = "not_observed"

    @classmethod
    def of(cls, value: bool | None) -> "Observation":
        """Lift an ``Optional[bool]`` into the tri-state. ``None`` means the check
        did not run — it does NOT mean False."""
        if value is None:
            return cls.UNKNOWN
        return cls.TRUE if value else cls.FALSE

    @classmethod
    def coerce(cls, value) -> "Observation":
        """Accept a member, a wire string, or a legacy ``Optional[bool]``."""
        if isinstance(value, cls):
            return value
        if isinstance(value, bool) or value is None:
            return cls.of(value)
        try:
            return cls(str(value))
        except ValueError:
            return cls.UNKNOWN

    @property
    def is_known(self) -> bool:
        return self is not Observation.UNKNOWN

    def as_optional_bool(self) -> bool | None:
        """For callers that still need ``Optional[bool]`` — UNKNOWN stays None."""
        if self is Observation.UNKNOWN:
            return None
        return self is Observation.TRUE

    def label(self, yes: str, no: str, unknown: str = "Not observed") -> str:
        """Operator-facing wording. `unknown` deliberately reads as a statement of
        what the camera saw, not as a detector failure ("not detected")."""
        return {Observation.TRUE: yes, Observation.FALSE: no}.get(self, unknown)


class BailsState(str, Enum):
    """Whether the wicket was broken. A domain-named tri-state: the vocabulary an
    umpire uses ("dislodged" / "intact") beats a generic true/false at the point
    of use, while carrying the same three-outcome guarantee."""

    DISLODGED = "dislodged"
    INTACT = "intact"
    NOT_OBSERVED = "not_observed"

    @classmethod
    def coerce(cls, value) -> "BailsState":
        """Legacy payloads used ``None`` for "no detector". That is UNKNOWN, and
        must never be read as INTACT."""
        if isinstance(value, cls):
            return value
        if value is None:
            return cls.NOT_OBSERVED
        try:
            return cls(str(value).lower())
        except ValueError:
            return cls.NOT_OBSERVED

    @property
    def is_known(self) -> bool:
        return self is not BailsState.NOT_OBSERVED

    @property
    def wicket_broken(self) -> bool | None:
        """None when unobserved — callers must not treat that as "not broken"."""
        if self is BailsState.NOT_OBSERVED:
            return None
        return self is BailsState.DISLODGED

    def label(self) -> str:
        return {
            BailsState.DISLODGED: "Dislodged",
            BailsState.INTACT: "Intact",
        }.get(self, "Not observed")
